File: packages/oxlint-plugin-react-doctor/src/plugin/rules/react-builtins/jsx-no-duplicate-props.ts
import { defineRule } from "../../utils/define-rule.js";
import type { EsTreeNodeOfType } from "../../utils/es-tree-node-of-type.js";
import { isNodeOfType } from "../../utils/is-node-of-type.js";
import type { Rule } from "../../utils/rule.js";

// Port of `oxc_linter::rules::react::jsx_no_duplicate_props`. Walks each
// JSXOpeningElement's attribute list and reports any `JSXAttribute` whose
// identifier name has already been seen in the same opening tag. Spread
// attributes (`{...props}`) are skipped — they don't carry a static name.
// OXC's port intentionally skips `ignoreCase` (cases differ → distinct).
export const jsxNoDuplicateProps = defineRule<Rule>({
  id: "jsx-no-duplicate-props",
  title: "Duplicate props on element",
  severity: "error",
  recommendation:
    "Remove or rename one of the duplicate props so the later value does not silently override the earlier one.",
  create: (context) => ({
    JSXOpeningElement(node: EsTreeNodeOfType<"JSXOpeningElement">) {
      const seenPropNames = new Set<string>();
      for (const attribute of node.attributes) {
        if (!isNodeOfType(attribute, "JSXAttribute")) continue;
        if (!isNodeOfType(attribute.name, "JSXIdentifier")) continue;
        const propName = attribute.name.name;
        if (seenPropNames.has(propName)) {
          context.report({
            node: attribute,
            message: `Your users can get the wrong value because React keeps only the last "${propName}" & drops the first.`,
          });
        }
        seenPropNames.add(propName);
      }
    },
  }),
});

File: native/oxlint/rules/jsx-no-duplicate-props.rs
use oxc_ast::{
    AstKind,
    ast::{JSXAttributeItem, JSXAttributeName},
};
use oxc_diagnostics::OxcDiagnostic;
use oxc_macros::declare_oxc_lint;
use rustc_hash::FxHashSet;

use crate::{AstNode, context::LintContext, rule::Rule};

#[derive(Debug, Default, Clone)]
pub struct JsxNoDuplicateProps;

declare_oxc_lint!(
    /// Disallow duplicate JSX props.
    JsxNoDuplicateProps,
    react_doctor_native,
    correctness,
    version = "0.1.0",
    short_description = "Disallow duplicate JSX props.",
);

impl Rule for JsxNoDuplicateProps {
    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
        let AstKind::JSXOpeningElement(opening_element) = node.kind() else {
            return;
        };
        let mut seen_prop_names = FxHashSet::default();
        for attribute in &opening_element.attributes {
            let JSXAttributeItem::Attribute(attribute) = attribute else {
                continue;
            };
            let JSXAttributeName::Identifier(identifier) = &attribute.name else {
                continue;
            };
            if !seen_prop_names.insert(identifier.name) {
                ctx.diagnostic(
                    OxcDiagnostic::warn(format!(
                        "Your users can get the wrong value because React keeps only the last \"{}\" & drops the first.",
                        identifier.name
                    ))
                    .with_label(attribute.span),
                );
            }
        }
    }
}
